### PR TITLE
capitalize first letter of search categories by default

### PR DESCRIPTION
**Pull request by WMDE-Fisch**
see https://phabricator.wikimedia.org/T117559

### CODE REVIEW

File: DeepCat.js [hunk 1]
```diff
@@ -7,7 +7,7 @@
  */
 ( function( $, mw ) {
 	var DeepCat = {},
-		keyString = 'deepcat:',
+		keyString = 'test:',
```
Comment (thiemowmde) [inline DeepCat.js]: Huh?

**Comment by WMDE-Fisch** [comment]
ups

**Comment by WMDE-Fisch** [comment]
fixed ;-)

**Comment by thiemowmde** [comment]
+2

**Comment by wmde-manicki** [comment]
TL;DR: just skip to the last paragraph where I suggest some possible "improvement" to this feature which is already working pretty good.

Tried this out and works as it is intended but I'd still have some remarks (do not mean that this couldn't be merged already, just the old man's rant mode on). Sorry for the lengthy read.

Remark 1. https://phabricator.wikimedia.org/T117559 is specifically about the first character of the category name, and so does the trick used here. That works perfect n German Wikipedia with `deepcat:physik`. Also works with `deepcat:"gemeinde in Deutschland"`. But if user assumed that all characters in the category name could be entered lower-case and searched for `deepcat:"gemeinde in Deutschland"` then there will no results.

It is a property of catgraph which is traversing the graph where nodes are the actual page (ie. category in this case) names. That is good and I think there is no need to consider changing this (even if there were technical possibilities to do so, although I suspect it might be rather impossible to do).
While I still think capitalizing the very first word in the category is fine and might be helpful for users, and I am not even sure if the [user's report](https://de.wikipedia.org/wiki/Wikipedia:Technische_W%C3%BCnsche/Fehlermeldungen#Gro.C3.9Fschreibung_des_Kategoriennamens_n.C3.B6tig) has concerned multi-word category names, I believe it might be worth mentioning it clear in the docs that search terms used with the `deepcat` keyword should use the actual category name used in the wiki, including capitalization of words. We might be capitalizing first word for user's convenience but it should be a general recommendation to use exact category names. Unless it is already said in docs - sorry if I am missing something.

Remark 2. A solution used here is based on an assumption that category names (same as page names in Mediawiki in general) must start with capital letter (if they are written in Latin script). This is generally true but it could be also changed in wiki settings, see: https://meta.wikimedia.org/wiki/Help:Page_name#Case-sensitivity_of_the_first_character and https://www.mediawiki.org/wiki/Manual:$wgCapitalLinks. 

It looks like non-capitalized categories are only used in Wiktionaries (which distincts between words written in lower and upper case) and Lojban Wikipedia (Lojban seems to only use lower-case letters?). See: https://phabricator.wikimedia.org/diffusion/OMWC/browse/master/wmf-config/InitialiseSettings.php;0a9e06f637d66a177d142a572f14f1bfa5e0141b$40). I'll focus on Wiktionary (with all respect to Lojban Wikipedia).
Usage of lower-cased categories varies between language versions of Wiktionary. For instance, English version has plenty of thousands of these, Polish couple of thousands, while Finnish has only some 10-20 such categories.
Forcing a gadget to always capitalize the category name would make it impossible to include things e.g. [en:Cats](https://en.wiktionary.org/wiki/Category:en:Cats) or [zh:Planets_of_the_Solar_System](https://en.wiktionary.org/wiki/Category:zh:Planets_of_the_Solar_System) in Deepcat searches on Wiktionary.
Still I have no really idea what is the usage of the gadget on Wiktionary. Was it even hardly used there I believe it shouldn't close the possibility to use lower-cased categories in search.

So finally, the conclusion: how about make it possible to switch off the capitalization of the first word by adding some option to the gadget or something like this? I believe leaving capitalization on by default would be fine given the majority of wikis only allows capitalized category names. And maybe make a clear suggestion in the documentation that exact category names incl. capitalization are preferred.

**Comment by WMDE-Fisch** [comment]
hmmm good points there. I tried to read the value of `$wgCapitalLinks` with `mw.config.get( 'wgCapitalLinks' )`  but got only `null` even on Lojban wiki.

Another solution would be case insensitive CatGraph request. But I dont know if thats possible. @jkroll20 ?

**Comment by wmde-manicki** [comment]
yes, if it was really intended to have case-insensitive searches with the gadget, then CatGraph should do such search. I guess it might be hard to have keeping the current performance level.

`mw.config` does not expose `wgCapitalLinks`: https://www.mediawiki.org/wiki/Manual:Interface/JavaScript#mw.config This could be probably changed but I am not sure if us needing it for the gadget would be a good enough justification :)

**Comment by wmde-manicki** [comment]
I'll merge this as it improves the gadget as least in some manner. I'd make a change that gives a possibility to turn the capitalization off (for wiktionary usage) as a follow up.
Question, how (if even possible) to have case-insensitive CatGraph searches remains open but it would be definitely better to discuss it somewhere else? Phabricator ticket maybe?

**Comment by kroll-j** [comment]
CatGraph itself only carries page_ids, not category titles. The lookup is done in catgraph-jsonp. [This commit](https://github.com/wmde/catgraph-jsonp/commit/22d3aa6e13492c65848f5a2951fd295a6568f3c6) does the following:
1. Capitalize the category title and return its ID if  found.
2. Look for the category title with original capitalization, and return its ID if found.
3. "Last resort": Do a case-insensitive SQL lookup and return the ID if found. 

Rationale: Case insensitive SQL search bypasses the SQL index, AFAIK. This is probably not a problem now, but it might matter when more people use the gadget. So I do 1) and 2) first, which uses the index and should catch 90% of cases.

So all queries to catgraph-jsonp are case insensitive now. No change to the gadget code should be necessary.

**Comment by wmde-manicki** [comment]
Sounds good, thanks! So now we only have to revert the change to the gadget :)

**Comment by WMDE-Fisch** [comment]
Wohooo! ( good we did not deploy the change yet )

**Comment by kroll-j** [comment]
Yes, the change to the gadget probably does not hurt, but it's better to do this in one place only, to avoid confusion. So I guess you should revert it.
